Dynamic UserAgent string

MarsianMan opened the pull request:
Injected version to manifest via Maven and modified user agent to use injected version rather than hardcoded value.

MarsianMan commented on libhoney/pom.xml:
```diff
@@ -62,6 +62,18 @@
                             implementation="org.apache.maven.plugins.shade.resource.ApacheLicenseResourceTransformer">
                         </transformer>
+                        <transformer
+                            implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
```
This is the Maven shade equivalent to "addDefaultSpecificationEntries" and "addDefaultImplementationEntries"[1] as found in beeline-java parent pom (reproduced below).

```xml
                            <!-- Add implementation/specification version to manifest file so we can read it out at runtime
                            (e.g. [class].class.getPackage().getImplementationVersion()) -->
                            
                            <addDefaultImplementationEntries>true</addDefaultImplementationEntries>
                            <addDefaultSpecificationEntries>true</addDefaultSpecificationEntries>
```

[1] https://maven.apache.org/shared/maven-archiver/

MarsianMan commented on libhoney/src/test/java/io/honeycomb/libhoney/End2EndTest.java:
```diff
@@ -82,7 +82,7 @@ public void sendingOneSimpleEvent() throws InterruptedException {
             .withHeader("Content-Type", containing("application/json"))
             .withHeader("X-Honeycomb-Team", equalTo("testWriteKey"))
-            .withHeader("User-Agent", equalTo("libhoneycomb-java/1.0.0"))
+            .withHeader("User-Agent", containing("libhoneycomb-java/"))
```
Removed hard coded from test as the version will now vary based on dependency version.

MarsianMan commented on pom.xml:
```diff
@@ -78,7 +78,7 @@
         <reportsPluginVersion>2.9</reportsPluginVersion>
         <mavenSourcePluginVersion>3.0.1</mavenSourcePluginVersion>
-        <javadocPluginVersion>3.0.0</javadocPluginVersion>
+        <javadocPluginVersion>3.0.1</javadocPluginVersion>
```
Upgraded due to bug affecting certain versions of Java (not sure if these are versions are intended for support) 
https://stackoverflow.com/questions/49460751/execution-attach-javadocs-of-goal-org-apache-maven-pluginsmaven-javadoc-plugin

tomoNeoworks commented on libhoney/src/test/java/io/honeycomb/libhoney/transport/batch/impl/HoneycombBatchConsumerTest.java:
```diff
@@ -144,7 +144,9 @@ public void GIVEN_anAdditionalUserAgent_EXPECT_requestToContainUserAgentHeaderWi
-        assertThat(value.getFirstHeader("user-agent").getValue()).isEqualTo("libhoneycomb-java/1.0.0 beeline/1.0.0");
+        String userAgent = value.getFirstHeader("user-agent").getValue();
+        assertThat(userAgent).contains("libhoneycomb-java/");
+        assertThat(userAgent).contains("beeline/");
```
We can match on the entire additional user agent string here ("beeline/1.0.0"), including the version.

MarsianMan commented on libhoney/src/main/java/io/honeycomb/libhoney/transport/batch/impl/HoneycombBatchConsumer.java:
```diff
@@ -48,7 +49,12 @@
+     * creating an actual artifact.
+     */
+    private static final String USER_AGENT = "libhoneycomb-java/" +
+        (LibHoney.class.getPackage().getImplementationVersion()==null ? "1.0.0" : LibHoney.class.getPackage().getImplementationVersion());
```
Is there a better default version to signify this is coming from IDE rather than a built artifact? Just wondering if that differentiation is useful for analytics on Honeycomb's side.

tomoNeoworks replied:
I thought we might only inject this static version number in the unit tests (e.g. via a package-private constructor just for the purposes of testing). In the e2e tests it could still default to null, but at least we would have some code coverage. This is an alternative approach to having to put a default version in the actual source.

MarsianMan commented:
> I would consider whether we can't get more test coverage here; I assume that Honeycomb's backend is matching on the exact user agent format, so will break if this changes. Trying to increase the coverage in the End2EndTest is not the right way to go, but perhaps we could add a test on the exact format in the HoneycombBatchConsumerTest. This would require injecting the source of the Libhoney version via a protected constructor. As this project is JDK 7-compatible, you can't use a lambda so this won't look pretty.

I have updated the tests to use pattern based version checking which should be valid no matter the version. I have also added a default version to ensure that these checks do not fail with "null" if tests are run outside of a Maven build (e.g. IDE). I am not sure if this is what you intended?
